File: golang/yaml/encoder.go
package yaml

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/goccy/go-yaml"
)

type Encoder interface {
	Encode(*EncodeInput) *EncodeOutput
}

func NewEncoder() Encoder {
	return &encoder{}
}

type encoder struct{}

func (e *encoder) Encode(input *EncodeInput) *EncodeOutput {
	g := convertToGo(input.Value)
	switch input.Format {
	default:
		return &EncodeOutput{IsError: true, ErrorMessage: fmt.Sprintf("unexpected format %v", input.Format)}
	case EncodeFormat_ENCODE_FORMAT_JSON:
		b := bytes.NewBuffer(nil)
		e := json.NewEncoder(b)
		if input.Pretty {
			e.SetIndent("", "  ")
		}
		if err := e.Encode(g); err != nil {
			return &EncodeOutput{
				IsError:      true,
				ErrorMessage: fmt.Sprintf("fail to encode json: %+v", err),
			}
		}
		return &EncodeOutput{Result: b.String()}
	case EncodeFormat_ENCODE_FORMAT_YAML:
		b := bytes.NewBuffer(nil)
		e := yaml.NewEncoder(b)
		if input.Pretty {
			e = yaml.NewEncoder(b, yaml.Indent(2))
		}
		if err := e.Encode(g); err != nil {
			return &EncodeOutput{
				IsError:      true,
				ErrorMessage: fmt.Sprintf("fail to encode yaml: %+v", err),
			}
		}
		return &EncodeOutput{Result: b.String()}
	}
}

func convertToGo(v *Value) any {
	switch v.Type {
	default:
		panic(fmt.Sprintf("unexpected type %v", v.Type))
	case Type_TYPE_NULL:
		return nil
	case Type_TYPE_BOOL:
		return v.Bool
	case Type_TYPE_NUM:
		return v.Num
	case Type_TYPE_STR:
		return v.Str
	case Type_TYPE_ARR:
		arr := []any{}
		for _, elem := range v.Arr {
			arr = append(arr, convertToGo(elem))
		}
		return arr
	case Type_TYPE_OBJ:
		obj := map[string]any{}
		for key, value := range v.Obj {
			obj[key] = convertToGo(value)
		}
		return obj
	}
}

File: typescript/src/yaml/encoder.ts
import {EncodeInput, EncodeOutput, EncodeOutputSchema} from "./encoder_pb.js";
import {Type, Value, ValueSchema} from "./value_pb.js";
import {create} from "@bufbuild/protobuf";
import {stringify} from "yaml";

export class Encoder {
    encode(input: EncodeInput): EncodeOutput {
        try {
            const o = this.convertToJS(input.value ?? create(ValueSchema));
            return create(EncodeOutputSchema, {result: stringify(o)});
        } catch (e) {
            return create(EncodeOutputSchema, {
                isError: true,
                errorMessage: e instanceof Error ? e.message : JSON.stringify(e),
            });
        }
    }

    private convertToJS(v: Value): unknown {
        switch (v.type) {
            case Type.NULL:
                return null;
            case Type.BOOL:
                return v.bool;
            case Type.NUM:
                return v.num;
            case Type.STR:
                return v.str;
            case Type.ARR:
                return v.arr.map(this.convertToJS);
            case Type.OBJ:
                return Object.fromEntries(Object.entries(v.obj).map(([k, v]) => [k, this.convertToJS(v)]));
            default:
                throw new Error("unexpected value type: " + Type[v.type]);
        }
    }
}